src/Futhark/Optimise/ArrayShortCircuiting/LastUse.hs

```diff
@@ -45,41 +46,96 @@ type LastUseOp rep = Op (Aliases rep) -> Names -> LastUseM rep (LUTabFun, Names,
+data LastUseReader rep = LastUseReader
+  { onOp :: LastUseOp rep,
+    scope :: Scope (Aliases rep)
   }
 
 -- | Maps a variable or memory block to its aliases.
 type AliasTab = M.Map VName Names
 
-type LastUseM rep a = StateT AliasTab (Reader (LastUseReader rep)) a
+newtype LastUseM rep a = LastUseM (StateT AliasTab (Reader (LastUseReader rep)) a)
+  deriving
+    ( Monad,
+      Functor,
+      Applicative,
+      MonadReader (LastUseReader rep),
+      MonadState AliasTab
+    )
+
+instance HasScope (Aliases SeqMem) (LastUseM SeqMem) where
+  askScope = asks scope
+
+instance LocalScope (Aliases SeqMem) (LastUseM SeqMem) where
+  localScope sc (LastUseM m) = LastUseM $ do
+    local (\rd -> rd {scope = scope rd <> sc}) m
+
+instance HasScope (Aliases GPUMem) (LastUseM GPUMem) where
+  askScope = asks scope
+
+instance LocalScope (Aliases GPUMem) (LastUseM GPUMem) where
+  localScope sc (LastUseM m) = LastUseM $ do
+    local (\rd -> rd {scope = scope rd <> sc}) m
+
+runLastUseM :: LastUseOp rep -> LastUseM rep a -> a
+runLastUseM onOp (LastUseM m) =
+  runReader (evalStateT m mempty) (LastUseReader onOp mempty)
 
 aliasLookup :: VName -> LastUseM rep Names
 aliasLookup vname =
   gets $ fromMaybe mempty . M.lookup vname
 
-lastUseProg :: (CanBeAliased (Op rep), Mem rep inner) => LastUseOp rep -> Prog (Aliases rep) -> LUTabFun
-lastUseProg onOp prog =
-  runReader (evalStateT m mempty) (LastUseReader onOp)
-  where
-    m = do
-      let bound_in_consts =
-            progConsts prog
-              & concatMap (patNames . stmPat)
-              & namesFromList
-          consts = progConsts prog
-          funs = progFuns prog
-      (consts_lu, _) <- lastUseStms consts mempty mempty
-      (lus, _) <- mconcat <$> mapM (flip lastUseBody (mempty, bound_in_consts) . funDefBody) funs
-      pure $ consts_lu <> lus
+-- lastUseProg ::
+--   ( Mem rep inner,
+--     CanBeAliased (Op rep)
+--   ) =>
+--   LastUseOp rep ->
+--   Prog (Aliases rep) ->
+--   LUTabFun
+-- lastUseProg onOp prog =
+--   runLastUseM onOp (m prog)
+
+lastUseProg ::
+  ( LocalScope (Aliases rep) (LastUseM rep),
+    Mem rep inner,
+    CanBeAliased (Op rep),
+    HasMemBlock (Aliases rep)
+  ) =>
+  Prog (Aliases rep) ->
+  LastUseM rep LUTabFun
+lastUseProg prog =
+  let bound_in_consts =
+        progConsts prog
+          & concatMap (patNames . stmPat)
+          & namesFromList
+      consts = progConsts prog
+      funs = progFuns prog
+   in inScopeOf consts $ do
+        (consts_lu, _) <- lastUseStms consts mempty mempty
+        (lus, _) <- mconcat <$> mapM (lastUseFun bound_in_consts) funs
+        pure $ consts_lu <> lus
+
+lastUseFun ::
+  ( LocalScope (Aliases rep) (LastUseM rep),
+    Mem rep inner,
+    CanBeAliased (Op rep),
+    HasMemBlock (Aliases rep)
+  ) =>
+  Names ->
+  FunDef (Aliases rep) ->
+  LastUseM rep (LUTabFun, Names)
+lastUseFun bound_in_consts f =
+  inScopeOf f $
+    flip lastUseBody (mempty, bound_in_consts) $
+      funDefBody f
```

Comment:
`inSCopeOf f $ lastUseBody (funDefBody f) (mempty, bound_in_consts)`

It's not like this convoluted use of `flip` gets you out of using parens anyway.